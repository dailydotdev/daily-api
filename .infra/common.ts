export interface WorkerArgs {
  enableMessageOrdering?: boolean;
  ackDeadlineSeconds?: number;
  expirationPolicy?: {
    ttl: string;
  };
  deadLetterPolicy?: {
    deadLetterTopic: string;
    maxDeliveryAttempts: number;
  };
}

export interface Worker {
  topic: string;
  subscription: string;
  args?: WorkerArgs;
}

export const digestDeadLetter = 'api.v1.personalized-digest-email-dead-letter';

export const workers: Worker[] = [
  {
    topic: 'views',
    subscription: 'add-views-v2',
  },
  {
    topic: 'user-updated',
    subscription: 'api.user-updated-cio',
  },
  {
    topic: 'user-deleted',
    subscription: 'api.user-deleted-cio',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'comment-upvoted-rep',
  },
  {
    topic: 'comment-upvote-canceled',
    subscription: 'comment-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-scout-matched-slack-v2',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-freeform-images',
  },
  {
    topic: 'api.v1.post-content-edited',
    subscription: 'api.post-edited-freeform-images',
  },
  {
    topic: 'comment-commented',
    subscription: 'comment-commented-slack-message',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-rep',
  },
  {
    topic: 'post-upvote-canceled',
    subscription: 'post-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-downvoted',
    subscription: 'api.post-downvoted-rep',
  },
  {
    topic: 'api.v1.post-downvote-canceled',
    subscription: 'api.post-downvote-canceled-rep',
  },
  {
    topic: 'post-commented',
    subscription: 'api.post-commented-images',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-redis',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-slack-message',
  },
  {
    topic: 'post-commented',
    subscription: 'api.vordr-post-comment-prevented',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-redis',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'post-banned-rep',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.post-deleted-comments-cleanup',
  },
  {
    topic: 'pub-request',
    subscription: 'pub-request-rep',
  },
  {
    topic: 'username-changed',
    subscription: 'username-changed-api',
  },
  {
    topic: 'update-comments',
    subscription: 'update-comments-mention',
  },
  {
    topic: 'api.changes',
    subscription: 'api-cdc',
    args: { enableMessageOrdering: true },
  },
  {
    topic: 'api.v1.cdc-notifications',
    subscription: 'api.cdc-notifications',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-real-time',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-mail',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-push',
  },
  {
    topic: 'api.v1.source-privacy-updated',
    subscription: 'api.source-privacy-updated',
  },
  {
    topic: 'api.v1.content-image-deleted',
    subscription: 'api.delete-cloudinary-image',
  },
  {
    topic: 'api.v1.comment-edited',
    subscription: 'api.comment-edited-images',
  },
  {
    topic: 'api.v1.comment-deleted',
    subscription: 'api.comment-deleted-images',
  },
  // Notifications
  {
    topic: 'community-link-rejected',
    subscription: 'api.community-picks-failed-notification',
  },
  {
    topic: 'community-link-access',
    subscription: 'api.community-picks-granted-notification',
  },
  {
    topic: 'post-commented',
    subscription: 'api.article-new-comment-notification.post-commented',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.article-new-comment-notification.comment-commented',
  },
  {
    topic: 'api.v1.post-bookmark-reminder',
    subscription: 'api.post-bookmark-reminder-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-commented-images',
  },
  {
    topic: 'post-upvoted',
    subscription: 'api.article-upvote-milestone-notification',
  },
  {
    topic: 'send-analytics-report',
    subscription: 'api.article-analytics-notification',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.article-report-approved-notification',
  },
  {
    topic: 'pub-request',
    subscription: 'api.source-request-notification',
  },
  {
    topic: 'api.v1.new-post-mention',
    subscription: 'api.post-mention-notification',
  },
  {
    topic: 'api.v1.new-comment-mention',
    subscription: 'api.comment-mention-notification',
  },
  {
    topic: 'user-reputation-updated',
    subscription: 'api.user-reputation-updated-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-reply-notification',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'api.comment-upvote-milestone-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-notification-v2',
  },
  {
    topic: 'api.v1.member-joined-source',
    subscription: 'api.member-joined-source-notification',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-cio',
  },
  {
    topic: 'api.v1.banner-added',
    subscription: 'api.banner-added',
  },
  {
    topic: 'api.v1.banner-deleted',
    subscription: 'api.banner-deleted',
  },
  {
    topic: 'yggdrasil.v1.content-published',
    subscription: 'api.content-published',
  },
  {
    topic: 'api.v1.source-member-role-changed',
    subscription: 'api.source-member-role-changed-notification',
  },
  {
    topic: 'analytics-api.v1.experiment-allocated',
    subscription: 'api.experiment-allocated',
  },
  {
    topic: 'api.v1.source-created',
    subscription: 'api.source-created-squad-user-action',
  },
  {
    topic: 'api.v1.post-collection-updated',
    subscription: 'api.post-collection-updated-notification',
  },
  {
    topic: 'api.v1.user-readme-updated',
    subscription: 'api.user-readme-images',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-personalized-digest-send-type',
  },
  {
    topic: 'api.v1.comment-downvoted',
    subscription: 'api.comment-downvoted-rep',
  },
  {
    topic: 'api.v1.comment-downvote-canceled',
    subscription: 'api.comment-downvote-canceled-rep',
  },
  {
    topic: 'api.v1.squad-public-request',
    subscription: 'api.v1.squad-public-request-notification',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-updated-cio',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-reset-notification',
  },
  {
    topic: 'api.v1.squad-featured-updated',
    subscription: 'api.squad-featured-updated-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-slack-channel-send',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.vordr-post-prevented',
  },
  {
    topic: 'api.v1.user-company-approved',
    subscription: 'api.user-company-approved-cio',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-user-notification',
  },
  {
    topic: 'api.v1.user-top-reader',
    subscription: 'api.user-top-reader-added',
  }
];

export const personalizedDigestWorkers: Worker[] = [
  {
    topic: 'api.v1.generate-personalized-digest',
    subscription: 'api.personalized-digest-email',
  },
  {
    topic: digestDeadLetter,
    subscription: 'api.personalized-digest-email-dead-letter-log',
  },
];
