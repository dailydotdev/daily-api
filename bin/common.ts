import fastq from 'fastq';
import { ReadStream } from 'fs';

const QUEUE_CONCURRENCY = 1;

export const runInQueueStream = async <T>(
  stream: ReadStream,
  callback: (props: T) => Promise<void>,
  queue = QUEUE_CONCURRENCY,
) => {
  let insertCount = 0;
  const insertQueue = fastq.promise(async (props: T) => {
    await callback(props);

    insertCount += 1;
  }, queue);

  stream.on('data', insertQueue.push);

  await new Promise((resolve, reject) => {
    stream.on('error', reject);
    stream.on('end', resolve);
  });
  await insertQueue.drained();
  console.log('update finished with a total of: ', insertCount);
};
