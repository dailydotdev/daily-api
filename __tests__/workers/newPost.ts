import { Connection, getConnection } from 'typeorm';
import { FastifyInstance } from 'fastify';
import { SearchIndex } from 'algoliasearch';
import { mocked } from 'ts-jest/utils';
import Mock = jest.Mock;

import appFunc from '../../src/background';
import worker from '../../src/workers/newPost';
import { expectSuccessfulBackground, saveFixtures } from '../helpers';
import {
  Keyword,
  Post,
  PostKeyword,
  PostTag,
  Source,
  User,
} from '../../src/entity';
import { sourcesFixture } from '../fixture/source';
import { getPostsIndex, notifyPostAuthorMatched } from '../../src/common';

let con: Connection;
let app: FastifyInstance;
let saveObjectMock: Mock;

jest.mock('../../src/common/algolia', () => ({
  ...jest.requireActual('../../src/common/algolia'),
  getPostsIndex: jest.fn(),
}));

jest.mock('../../src/common', () => ({
  ...jest.requireActual('../../src/common'),
  notifyPostAuthorMatched: jest.fn(),
}));

beforeAll(async () => {
  con = await getConnection();
  app = await appFunc();
  return app.ready();
});

beforeEach(async () => {
  jest.resetAllMocks();
  saveObjectMock = jest.fn();
  mocked(getPostsIndex).mockReturnValue(({
    saveObject: saveObjectMock,
  } as unknown) as SearchIndex);

  await saveFixtures(con, Source, sourcesFixture);
});

it('should save a new post with basic information', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(saveObjectMock).toBeCalledWith({
    objectID: posts[0].id,
    title: 'Title',
    createdAt: expect.any(Number),
    views: 0,
    readTime: undefined,
    pubId: 'a',
    _tags: undefined,
  });
  const tags = await con.getRepository(PostTag).find();
  expect(tags.length).toEqual(0);
  expect(posts[0]).toMatchSnapshot({
    createdAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
  });
});

it('should save a new post with full information', async () => {
  const timestamp = new Date(2020, 5, 11, 1, 17);

  await con.getRepository(Keyword).save([
    { value: 'javascript', occurrences: 20, status: 'allow' },
    { value: 'html', occurrences: 15, status: 'allow' },
    { value: 'webdev', occurrences: 5 },
    { value: 'js', occurrences: 5, status: 'synonym', synonym: 'javascript' },
    { value: 'nodejs', occurrences: 5, status: 'deny' },
  ]);
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    publishedAt: timestamp.toISOString(),
    image: 'https://image.com',
    ratio: 2,
    placeholder: 'data:image/jpeg;base64,placeholder',
    tags: ['webdev', 'javascript', 'html'],
    keywords: ['webdev', 'javascript', 'html', 'js', 'nodejs'],
    siteTwitter: 'site',
    creatorTwitter: 'creator',
    readTime: '5.123',
  });
  expect(saveObjectMock).toBeCalledWith({
    objectID: expect.any(String),
    title: 'Title',
    createdAt: expect.any(Number),
    views: 0,
    readTime: 5,
    pubId: 'a',
    _tags: ['webdev', 'javascript', 'html'],
  });
  const posts = await con.getRepository(Post).find();
  const tags = await con.getRepository(PostTag).find({ select: ['tag'] });
  expect(posts.length).toEqual(1);
  expect(posts[0]).toMatchSnapshot({
    createdAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
  });
  expect(tags).toMatchSnapshot();
});

it('should handle empty tags array', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    tags: [],
  });
  expect(saveObjectMock).toBeCalledWith({
    objectID: expect.any(String),
    title: 'Title',
    createdAt: expect.any(Number),
    views: 0,
    readTime: undefined,
    pubId: 'a',
    _tags: [],
  });
  const posts = await con.getRepository(Post).find();
  const tags = await con.getRepository(PostTag).find();
  expect(posts.length).toEqual(1);
  expect(tags.length).toEqual(0);
});

it('should save keywords', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    keywords: ['vue', 'nodejs'],
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  const postKeywords = await con
    .getRepository(PostKeyword)
    .find({ select: ['keyword'], order: { keyword: 'ASC' } });
  const keywords = await con
    .getRepository(Keyword)
    .find({ select: ['value', 'status'], order: { value: 'ASC' } });
  expect(posts.length).toEqual(1);
  expect(postKeywords).toMatchSnapshot();
  expect(keywords).toMatchSnapshot();
});

it('should ignore numerical only keywords', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    keywords: ['vue', 'nodejs', '123'],
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  const postKeywords = await con
    .getRepository(PostKeyword)
    .find({ select: ['keyword'], order: { keyword: 'ASC' } });
  const keywords = await con
    .getRepository(Keyword)
    .find({ select: ['value', 'status'], order: { value: 'ASC' } });
  expect(posts.length).toEqual(1);
  expect(postKeywords).toMatchSnapshot();
  expect(keywords).toMatchSnapshot();
});

it('should increase occurrences by one when keyword exists', async () => {
  await con.getRepository(Keyword).save([{ value: 'nodejs', status: 'allow' }]);
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    keywords: ['vue', 'nodejs'],
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  const postKeywords = await con
    .getRepository(PostKeyword)
    .find({ select: ['keyword'], order: { keyword: 'ASC' } });
  const keywords = await con.getRepository(Keyword).find({
    select: ['value', 'status', 'occurrences'],
    order: { value: 'ASC' },
  });
  expect(posts.length).toEqual(1);
  expect(postKeywords).toMatchSnapshot();
  expect(keywords).toMatchSnapshot();
});

it('should handle duplicate keywords', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    keywords: ['vue', 'nodejs', 'vue'],
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  const postKeywords = await con
    .getRepository(PostKeyword)
    .find({ select: ['keyword'], order: { keyword: 'ASC' } });
  const keywords = await con
    .getRepository(Keyword)
    .find({ select: ['value', 'status'], order: { value: 'ASC' } });
  expect(posts.length).toEqual(1);
  expect(postKeywords).toMatchSnapshot();
  expect(keywords).toMatchSnapshot();
});

it('should replace synonym keywords', async () => {
  await con.getRepository(Keyword).save([
    { value: 'node', status: 'synonym', synonym: 'nodejs' },
    { value: 'nodejs', status: 'allow' },
  ]);
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    keywords: ['vue', 'node'],
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  const postKeywords = await con
    .getRepository(PostKeyword)
    .find({ select: ['keyword'], order: { keyword: 'ASC' } });
  const keywords = await con
    .getRepository(Keyword)
    .find({ select: ['value', 'status'], order: { value: 'ASC' } });
  expect(posts.length).toEqual(1);
  expect(postKeywords).toMatchSnapshot();
  expect(keywords).toMatchSnapshot();
});

it('should handle empty keywords array', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    keywords: [],
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  const postKeywords = await con.getRepository(PostKeyword).find();
  const keywords = await con.getRepository(Keyword).find();
  expect(posts.length).toEqual(1);
  expect(postKeywords.length).toEqual(0);
  expect(keywords.length).toEqual(0);
});

it('should ignore null value violation', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: null,
    publicationId: 'a',
  });
  expect(saveObjectMock).toBeCalledTimes(0);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(0);
});

it('should not save post with existing url', async () => {
  await con.getRepository(Post).save({
    id: 'p2',
    shortId: 'p2',
    title: 'Title 2',
    url: 'https://post.com',
    sourceId: 'b',
  });

  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
  });
  expect(saveObjectMock).toBeCalledTimes(0);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
});

it('should not save post when url matches existing canonical url', async () => {
  await con.getRepository(Post).save({
    id: 'p2',
    shortId: 'p2',
    title: 'Title 2',
    url: 'https://post.com',
    canonicalUrl: 'https://post.dev',
    sourceId: 'b',
  });

  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.dev',
    publicationId: 'a',
  });
  expect(saveObjectMock).toBeCalledTimes(0);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
});

it('should not save post when canonical url matches existing url', async () => {
  await con.getRepository(Post).save({
    id: 'p2',
    shortId: 'p2',
    title: 'Title 2',
    url: 'https://post.com',
    canonicalUrl: 'https://post.dev',
    sourceId: 'b',
  });

  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.io',
    canonicalUrl: 'https://post.com',
    publicationId: 'a',
  });
  expect(saveObjectMock).toBeCalledTimes(0);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
});

it('should not save post when canonical url matches existing canonical url', async () => {
  await con.getRepository(Post).save({
    id: 'p2',
    shortId: 'p2',
    title: 'Title 2',
    url: 'https://post.com',
    canonicalUrl: 'https://post.dev',
    sourceId: 'b',
  });

  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.io',
    canonicalUrl: 'https://post.dev',
    publicationId: 'a',
  });
  expect(saveObjectMock).toBeCalledTimes(0);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
});

it('should match post to author', async () => {
  await con.getRepository(User).save([
    {
      id: '1',
      name: 'Ido',
      image: 'https://daily.dev/ido.jpg',
      twitter: 'idoshamun',
    },
  ]);

  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    creatorTwitter: '@idoshamun',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(saveObjectMock).toBeCalledWith({
    objectID: posts[0].id,
    title: 'Title',
    createdAt: expect.any(Number),
    views: 0,
    readTime: undefined,
    pubId: 'a',
    _tags: undefined,
  });
  expect(posts[0]).toMatchSnapshot({
    createdAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
  });
  expect(mocked(notifyPostAuthorMatched).mock.calls[0].slice(1)).toEqual([
    posts[0].id,
    '1',
  ]);
});

it('should match post to author based on username', async () => {
  await con.getRepository(User).save([
    {
      id: '1',
      name: 'Ido',
      image: 'https://daily.dev/ido.jpg',
      username: 'idoshamun',
    },
  ]);
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    creatorTwitter: '@idoshamun',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(saveObjectMock).toBeCalledWith({
    objectID: posts[0].id,
    title: 'Title',
    createdAt: expect.any(Number),
    views: 0,
    readTime: undefined,
    pubId: 'a',
    _tags: undefined,
  });
  expect(posts[0]).toMatchSnapshot({
    createdAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
  });
  expect(mocked(notifyPostAuthorMatched).mock.calls[0].slice(1)).toEqual([
    posts[0].id,
    '1',
  ]);
});

it('should not match post to author', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    creatorTwitter: '@nouser',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(saveObjectMock).toBeCalledWith({
    objectID: posts[0].id,
    title: 'Title',
    createdAt: expect.any(Number),
    views: 0,
    readTime: undefined,
    pubId: 'a',
    _tags: undefined,
  });
  expect(posts[0]).toMatchSnapshot({
    createdAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
  });
  expect(notifyPostAuthorMatched).toBeCalledTimes(0);
});

it('should clear empty creator twitter', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    creatorTwitter: '',
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(posts[0].creatorTwitter).toBeNull();
});

it('should clear invalid creator twitter', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.com',
    publicationId: 'a',
    creatorTwitter: '@',
  });
  expect(saveObjectMock).toBeCalledTimes(1);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(posts[0].creatorTwitter).toBeNull();
});

it('should not save post when author is banned', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Title',
    url: 'https://post.io',
    publicationId: 'a',
    creatorTwitter: '@NewGenDeveloper',
  });
  expect(saveObjectMock).toBeCalledTimes(0);
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(0);
});

it('should unescape html text', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'It&#039;s ok jQuery, I still love you',
    url: 'https://post.com',
    publicationId: 'a',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(saveObjectMock).toBeCalledWith({
    objectID: posts[0].id,
    title: `It's ok jQuery, I still love you`,
    createdAt: expect.any(Number),
    views: 0,
    readTime: undefined,
    pubId: 'a',
    _tags: undefined,
  });
  const tags = await con.getRepository(PostTag).find();
  expect(tags.length).toEqual(0);
  expect(posts[0]).toMatchSnapshot({
    createdAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
  });
});

it('should keep html-like text', async () => {
  await expectSuccessfulBackground(app, worker, {
    id: 'p1',
    title: 'Here is my <progress> element',
    url: 'https://post.com',
    publicationId: 'a',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(1);
  expect(saveObjectMock).toBeCalledWith({
    objectID: posts[0].id,
    title: 'Here is my <progress> element',
    createdAt: expect.any(Number),
    views: 0,
    readTime: undefined,
    pubId: 'a',
    _tags: undefined,
  });
  const tags = await con.getRepository(PostTag).find();
  expect(tags.length).toEqual(0);
  expect(posts[0]).toMatchSnapshot({
    createdAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
  });
});
