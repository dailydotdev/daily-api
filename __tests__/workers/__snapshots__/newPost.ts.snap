// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle duplicate keywords 1`] = `
Array [
  PostKeyword {
    "keyword": "nodejs",
  },
  PostKeyword {
    "keyword": "vue",
  },
]
`;

exports[`should handle duplicate keywords 2`] = `
Array [
  Keyword {
    "status": "pending",
    "value": "nodejs",
  },
  Keyword {
    "status": "pending",
    "value": "vue",
  },
]
`;

exports[`should ignore numerical only keywords 1`] = `
Array [
  PostKeyword {
    "keyword": "nodejs",
  },
  PostKeyword {
    "keyword": "vue",
  },
]
`;

exports[`should ignore numerical only keywords 2`] = `
Array [
  Keyword {
    "status": "pending",
    "value": "nodejs",
  },
  Keyword {
    "status": "pending",
    "value": "vue",
  },
]
`;

exports[`should increase occurrences by one when keyword exists 1`] = `
Array [
  PostKeyword {
    "keyword": "nodejs",
  },
  PostKeyword {
    "keyword": "vue",
  },
]
`;

exports[`should increase occurrences by one when keyword exists 2`] = `
Array [
  Keyword {
    "occurrences": 2,
    "status": "allow",
    "value": "nodejs",
  },
  Keyword {
    "occurrences": 1,
    "status": "pending",
    "value": "vue",
  },
]
`;

exports[`should keep html-like text 1`] = `
Object {
  "authorId": null,
  "banned": false,
  "canonicalUrl": null,
  "comments": 0,
  "createdAt": Any<Date>,
  "creatorTwitter": null,
  "discussionScore": null,
  "id": Any<String>,
  "image": null,
  "lastTrending": null,
  "placeholder": null,
  "publishedAt": null,
  "ratio": null,
  "readTime": null,
  "score": Any<Number>,
  "sentAnalyticsReport": true,
  "shortId": Any<String>,
  "siteTwitter": null,
  "sourceId": "a",
  "tagsStr": null,
  "title": "Here is my <progress> element",
  "trending": null,
  "tweeted": false,
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
}
`;

exports[`should match post to author 1`] = `
Object {
  "authorId": "1",
  "banned": false,
  "canonicalUrl": null,
  "comments": 0,
  "createdAt": Any<Date>,
  "creatorTwitter": "@Idoshamun",
  "discussionScore": null,
  "id": Any<String>,
  "image": null,
  "lastTrending": null,
  "placeholder": null,
  "publishedAt": null,
  "ratio": null,
  "readTime": null,
  "score": Any<Number>,
  "sentAnalyticsReport": false,
  "shortId": Any<String>,
  "siteTwitter": null,
  "sourceId": "a",
  "tagsStr": null,
  "title": "Title",
  "trending": null,
  "tweeted": false,
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
}
`;

exports[`should match post to author based on username 1`] = `
Object {
  "authorId": "1",
  "banned": false,
  "canonicalUrl": null,
  "comments": 0,
  "createdAt": Any<Date>,
  "creatorTwitter": "@idoshamun",
  "discussionScore": null,
  "id": Any<String>,
  "image": null,
  "lastTrending": null,
  "placeholder": null,
  "publishedAt": null,
  "ratio": null,
  "readTime": null,
  "score": Any<Number>,
  "sentAnalyticsReport": false,
  "shortId": Any<String>,
  "siteTwitter": null,
  "sourceId": "a",
  "tagsStr": null,
  "title": "Title",
  "trending": null,
  "tweeted": false,
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
}
`;

exports[`should not match post to author 1`] = `
Object {
  "authorId": null,
  "banned": false,
  "canonicalUrl": null,
  "comments": 0,
  "createdAt": Any<Date>,
  "creatorTwitter": "@nouser",
  "discussionScore": null,
  "id": Any<String>,
  "image": null,
  "lastTrending": null,
  "placeholder": null,
  "publishedAt": null,
  "ratio": null,
  "readTime": null,
  "score": Any<Number>,
  "sentAnalyticsReport": true,
  "shortId": Any<String>,
  "siteTwitter": null,
  "sourceId": "a",
  "tagsStr": null,
  "title": "Title",
  "trending": null,
  "tweeted": false,
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
}
`;

exports[`should replace synonym keywords 1`] = `
Array [
  PostKeyword {
    "keyword": "nodejs",
  },
  PostKeyword {
    "keyword": "vue",
  },
]
`;

exports[`should replace synonym keywords 2`] = `
Array [
  Keyword {
    "status": "synonym",
    "value": "node",
  },
  Keyword {
    "status": "allow",
    "value": "nodejs",
  },
  Keyword {
    "status": "pending",
    "value": "vue",
  },
]
`;

exports[`should save a new post with basic information 1`] = `
Object {
  "authorId": null,
  "banned": false,
  "canonicalUrl": null,
  "comments": 0,
  "createdAt": Any<Date>,
  "creatorTwitter": null,
  "discussionScore": null,
  "id": Any<String>,
  "image": null,
  "lastTrending": null,
  "placeholder": null,
  "publishedAt": null,
  "ratio": null,
  "readTime": null,
  "score": Any<Number>,
  "sentAnalyticsReport": true,
  "shortId": Any<String>,
  "siteTwitter": null,
  "sourceId": "a",
  "tagsStr": null,
  "title": "Title",
  "trending": null,
  "tweeted": false,
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
}
`;

exports[`should save a new post with full information 1`] = `
Object {
  "authorId": null,
  "banned": false,
  "canonicalUrl": null,
  "comments": 0,
  "createdAt": Any<Date>,
  "creatorTwitter": "creator",
  "discussionScore": null,
  "id": Any<String>,
  "image": "https://image.com",
  "lastTrending": null,
  "placeholder": "data:image/jpeg;base64,placeholder",
  "publishedAt": 2020-06-11T01:17:00.000Z,
  "ratio": 2,
  "readTime": 5,
  "score": Any<Number>,
  "sentAnalyticsReport": true,
  "shortId": Any<String>,
  "siteTwitter": "site",
  "sourceId": "a",
  "tagsStr": "javascript,html",
  "title": "Title",
  "trending": null,
  "tweeted": false,
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
}
`;

exports[`should save a new post with full information 2`] = `
Array [
  PostTag {
    "tag": "webdev",
  },
  PostTag {
    "tag": "javascript",
  },
  PostTag {
    "tag": "html",
  },
]
`;

exports[`should save keywords 1`] = `
Array [
  PostKeyword {
    "keyword": "nodejs",
  },
  PostKeyword {
    "keyword": "vue",
  },
]
`;

exports[`should save keywords 2`] = `
Array [
  Keyword {
    "status": "pending",
    "value": "nodejs",
  },
  Keyword {
    "status": "pending",
    "value": "vue",
  },
]
`;

exports[`should unescape html text 1`] = `
Object {
  "authorId": null,
  "banned": false,
  "canonicalUrl": null,
  "comments": 0,
  "createdAt": Any<Date>,
  "creatorTwitter": null,
  "discussionScore": null,
  "id": Any<String>,
  "image": null,
  "lastTrending": null,
  "placeholder": null,
  "publishedAt": null,
  "ratio": null,
  "readTime": null,
  "score": Any<Number>,
  "sentAnalyticsReport": true,
  "shortId": Any<String>,
  "siteTwitter": null,
  "sourceId": "a",
  "tagsStr": null,
  "title": "It's ok jQuery, I still love you",
  "trending": null,
  "tweeted": false,
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
}
`;
